From b55a4be7b3a4e77eac6a2285cc0e048da7c68922 Mon Sep 17 00:00:00 2001
From: Pantelis Antoniou <panto@antoniou-consulting.com>
Date: Sat, 13 Oct 2012 16:37:24 +0300
Subject: [PATCH 11/11] ti_tscadc: Update with IIO map interface & deal with
 partial activation

Add an IIO map interface that consumers can use.
While we're here fix the mfd device in the case where a subdevice
might not be activated.
---
 drivers/iio/adc/ti_adc.c      |   49 +++++++++++++++++++++++++++++++++++------
 drivers/mfd/ti_tscadc.c       |   31 ++++++++++++++++++-------
 include/linux/mfd/ti_tscadc.h |    8 ++----
 3 files changed, 67 insertions(+), 21 deletions(-)

diff --git a/drivers/iio/adc/ti_adc.c b/drivers/iio/adc/ti_adc.c
index 14601f0..fa55bc4 100644
--- a/drivers/iio/adc/ti_adc.c
+++ b/drivers/iio/adc/ti_adc.c
@@ -20,9 +20,11 @@
 #include <linux/slab.h>
 #include <linux/interrupt.h>
 #include <linux/platform_device.h>
-#include <linux/io.h>
 #include <linux/iio/iio.h>
+#include <linux/iio/machine.h>
+#include <linux/iio/driver.h>
 
+#include <asm/io.h>
 #include <linux/mfd/ti_tscadc.h>
 #include <linux/platform_data/ti_adc.h>
 
@@ -30,6 +32,8 @@ struct adc_device {
 	struct ti_tscadc_dev *mfd_tscadc;
 	struct iio_dev *idev;
 	int channels;
+	char *buf;
+	struct iio_map *map;
 };
 
 static unsigned int adc_readl(struct adc_device *adc, unsigned int reg)
@@ -76,20 +80,28 @@ static void adc_step_config(struct adc_device *adc_dev)
 static int tiadc_channel_init(struct iio_dev *idev, struct adc_device *adc_dev)
 {
 	struct iio_chan_spec *chan_array;
-	int i;
+	struct iio_chan_spec *chan;
+	char *s;
+	int i, len, size, ret;
 
 	idev->num_channels = adc_dev->channels;
-	chan_array = kcalloc(idev->num_channels, sizeof(struct iio_chan_spec),
-					GFP_KERNEL);
 
+	size = idev->num_channels * (sizeof(struct iio_chan_spec) + 6);
+	chan_array = kzalloc(size, GFP_KERNEL);
 	if (chan_array == NULL)
 		return -ENOMEM;
 
-	for (i = 0; i < (idev->num_channels); i++) {
-		struct iio_chan_spec *chan = chan_array + i;
+	/* buffer space is after the array */
+	s = (char *)(chan_array + idev->num_channels);
+	chan = chan_array;
+	for (i = 0; i < idev->num_channels; i++, chan++, s += len + 1) {
+
+		len = sprintf(s, "AIN%d", i);
+
 		chan->type = IIO_VOLTAGE;
 		chan->indexed = 1;
 		chan->channel = i;
+		chan->datasheet_name = s;
 		chan->scan_type.sign = 'u';
 		chan->scan_type.realbits = 12;
 		chan->scan_type.storagebits = 32;
@@ -97,6 +109,30 @@ static int tiadc_channel_init(struct iio_dev *idev, struct adc_device *adc_dev)
 	}
 
 	idev->channels = chan_array;
+
+	size = (idev->num_channels + 1) * sizeof(struct iio_map);
+	adc_dev->map = kzalloc(size, GFP_KERNEL);
+	if (adc_dev->map == NULL) {
+		kfree(chan_array);
+		return -ENOMEM;
+	}
+
+	for (i = 0; i < idev->num_channels; i++) {
+		adc_dev->map[i].adc_channel_label = chan_array[i].datasheet_name;
+		adc_dev->map[i].consumer_dev_name = "any";
+		adc_dev->map[i].consumer_channel = chan_array[i].datasheet_name;
+	}
+	adc_dev->map[i].adc_channel_label = NULL;
+	adc_dev->map[i].consumer_dev_name = NULL;
+	adc_dev->map[i].consumer_channel = NULL;
+
+	ret = iio_map_array_register(idev, adc_dev->map);
+	if (ret != 0) {
+		kfree(adc_dev->map);
+		kfree(chan_array);
+		return -ENOMEM;
+	}
+
 	return idev->num_channels;
 }
 
@@ -175,7 +211,6 @@ static int __devinit tiadc_probe(struct platform_device *pdev)
 	platform_set_drvdata(pdev, idev);
 
 	return 0;
-
 err_unregister:
 	tiadc_channel_remove(idev);
 err_cleanup_channels:
diff --git a/drivers/mfd/ti_tscadc.c b/drivers/mfd/ti_tscadc.c
index bfb67bd..228f84a 100644
--- a/drivers/mfd/ti_tscadc.c
+++ b/drivers/mfd/ti_tscadc.c
@@ -153,25 +153,38 @@ static	int __devinit ti_tscadc_probe(struct platform_device *pdev)
 	ctrl |= CNTRLREG_TSCSSENB;
 	tscadc_writel(tscadc, REG_CTRL, ctrl);
 
+	tscadc->used_cells = 0;
+	tscadc->tsc_cell = -1;
+	tscadc->adc_cell = -1;
+
 	/* TSC Cell */
-	cell = &tscadc->cells[TSC_CELL];
-	cell->name = "tsc";
-	cell->platform_data = tscadc;
-	cell->pdata_size = sizeof(*tscadc);
+	if (tsc_wires > 0) {
+		tscadc->tsc_cell = tscadc->used_cells;
+		cell = &tscadc->cells[tscadc->used_cells++];
+		cell->name = "tsc";
+		cell->platform_data = tscadc;
+		cell->pdata_size = sizeof(*tscadc);
+	}
 
 	/* ADC Cell */
-	cell = &tscadc->cells[ADC_CELL];
-	cell->name = "tiadc";
-	cell->platform_data = tscadc;
-	cell->pdata_size = sizeof(*tscadc);
+	if (adc_channels > 0) {
+		tscadc->adc_cell = tscadc->used_cells;
+		cell = &tscadc->cells[tscadc->used_cells++];
+		cell->name = "tiadc";
+		cell->platform_data = tscadc;
+		cell->pdata_size = sizeof(*tscadc);
+	}
 
 	err = mfd_add_devices(&pdev->dev, pdev->id, tscadc->cells,
-			TSCADC_CELLS, NULL, 0, NULL);
+			tscadc->used_cells, NULL, 0, NULL);
 	if (err < 0)
 		goto err_disable_clk;
 
 	device_init_wakeup(&pdev->dev, true);
 	platform_set_drvdata(pdev, tscadc);
+
+	dev_info(&pdev->dev, "Probing in %s completed OK.\n", __func__);
+
 	return 0;
 
 err_disable_clk:
diff --git a/include/linux/mfd/ti_tscadc.h b/include/linux/mfd/ti_tscadc.h
index e0a6437..6e57433 100644
--- a/include/linux/mfd/ti_tscadc.h
+++ b/include/linux/mfd/ti_tscadc.h
@@ -125,11 +125,6 @@
 
 #define TSCADC_CELLS		2
 
-enum tscadc_cells {
-	TSC_CELL,
-	ADC_CELL,
-};
-
 struct mfd_tscadc_board {
 	struct tsc_data *tsc_init;
 	struct adc_data *adc_init;
@@ -139,6 +134,9 @@ struct ti_tscadc_dev {
 	struct device *dev;
 	void __iomem *tscadc_base;
 	int irq;
+	int used_cells;	/* 0-2 */
+	int tsc_cell;	/* -1 if not used */
+	int adc_cell;	/* -1 if not used */
 	struct mfd_cell cells[TSCADC_CELLS];
 
 	/* tsc device */
-- 
1.7.7.6

